source/language/openpulse.rst

```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be be
+  defined in the rotating frame (i.e. without the carrier) with the effect being
+  an equivalent application in the lab frame (i.e. with the carrier supplied by the ``frame``).
+  Another motivation is to more naturally implement a "virtual Z-gate", which does not require a
+  physical pulse but rather shifts the phase of all future pulses on that frame.
 
-Restrictions on the use of frames
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The frame is composed of three parts:
 
-There are two important considersations when dealing with frames.
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time of type ``dt`` which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``.
 
-First, frames are compile time constants. Runtime variables cannot be used as
-arguments to frames; the frame must be resolvable at compile time. This also
-means that assigning a frame to an alias (for the purpose of avoiding typing)
-needs to be done with ``const`` not ``let``.
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
 
-.. code-block: none
+Frame Construction
+~~~~~~~~~~~~~~~~~~
 
-   const driveframe = frameof("drive", %0);
+Frames are purely virtual and can be constructed using the ``Frame`` command
 
-Second, frames return references not values. This means that calling the same
-``frameof`` command in different places is the same as assigning the frame to a
-constant and referencing it multiple times.
+.. code-block:: javascript
 
-Frame manipulation
-~~~~~~~~~~~~~~~~~~
+  frame driveframe = Frame(5e9, 0.0);
 
-The (frequency, phase, scale) of a frame can be manipulated throughout program
-by referencing ``.frequency``, ``.phase``, and ``.scale``. Operations must be 
-appropriate for the respective type, ``float`` for frequency, ``angle`` for
-phase, and ``float`` for scale. Again, the exact precision of these calculations
-is hardware specific.
+When instantiated, the frame time starts at 0. ``Frame``s can also be copied using the
+``copy`` command with argument replacement
 
-Here's an example of manipulating the phase to calibrate an ``rz`` gate:
+.. code-block:: javascript
 
-.. code-block:: none
+  frame driveframe1 = Frame(1.0, 5e9, 0.0);
+  frame driveframe2 = copy(driveframe1, phase=driveframe1.phase + pi/2);
 
-   // Shift phase of qubit 0's "drive" frame by pi/4, eg. an rz gate with angle -pi/4
-   frameof("drive", %0).phase += pi/4;
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` (i.e. with
```

Comment:
Is this independent of the derived frame from this point forward?

Reply:
No; there are no dependencies between frames.

Reply:
Took me a few seconds to parse, I think you mean "yes; there are no dependencies between frames"

Reply:
The parenthetical "i.e. with the same frequency and time as driveframe1" might just make this more confusing. Just saying it's a copy might be enough.

Reply:
Would it be beneficial to define the copied frame w/ no aggregated phase and initial time = 0? This seems to be the main reason to copy a `frame`. For instance, if you have already used the `frame` and want a fresh one to use on a new physical channel. If you wanted the same time, you could just keep using the original frame.

Reply:
> The parenthetical "i.e. with the same frequency and time as driveframe1" might just make this more confusing. Just saying it's a copy might be enough.

Fixed.

> Would it be beneficial to define the copied frame w/ no aggregated phase and initial time = 0? This seems to be the main reason to copy a frame. For instance, if you have already used the frame and want a fresh one to use on a new physical channel. If you wanted the same time, you could just keep using the original frame.

You can do that by initializing a new frame via the `Frame(freq, phase)` initialize. The copy is there for the explicit cases where you want to same time as same other frame.